src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/InstructionEncoder.cs

```diff
@@ -401,23 +420,49 @@ public LabelHandle DefineLabel()
         public void Branch(ILOpCode code, LabelHandle label)
         {
             // throws if code is not a branch:
-            int size = code.GetBranchOperandSize();
+            int operandSize = code.GetBranchOperandSize();
+            // We want the offset before we add the opcode.
+            int ilOffset = Offset;
 
-            GetBranchBuilder().AddBranch(Offset, label, code);
             OpCode(code);
+            LabelOperand(code, label, operandSize, ilOffset);
+        }
 
-            // -1 points in the middle of the branch instruction and is thus invalid.
-            // We want to produce invalid IL so that if the caller doesn't patch the branches
-            // the branch instructions will be invalid in an obvious way.
-            if (size == 1)
-            {
-                CodeBuilder.WriteSByte(-1);
-            }
-            else
+        /// <summary>
+        /// Starts encoding a switch instruction.
+        /// </summary>
+        /// <param name="branchCount">The number of branches the instruction will have.</param>
+        /// <returns>A <see cref="SwitchInstructionEncoder"/> that will
+        /// be used to emit the labels for the branches.</returns>
+        /// <remarks>
+        /// Before using this <see cref="InstructionEncoder"/> in any other way,
+        /// the method <see cref="SwitchInstructionEncoder.Branch(LabelHandle)"/>
+        /// must be called on the returned value exactly <paramref name="branchCount"/>
+        /// times. Failure to do so will throw <see cref="InvalidOperationException"/>.
+        /// </remarks>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="branchCount"/>
+        /// less than or equal to zero.</exception>
+        public SwitchInstructionEncoder Switch(int branchCount)
+        {
+            if (branchCount <= 0)
             {
-                Debug.Assert(size == 4);
-                CodeBuilder.WriteInt32(-1);
+                Throw.ArgumentOutOfRange(nameof(branchCount));
             }
```

Comment:
Do we need to validate we're not already in a switch?

Reply:
It will happen a couple of lines later when we call `OpCode(ILOpCode.Switch);`.